internal/common/pid_sock_status.go

```diff
@@ -0,0 +1,83 @@
+package common
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/edulinq/autograder/internal/config"
+	"github.com/edulinq/autograder/internal/util"
+)
+
+const (
+	PID_FILENAME         = "autograder.pid"
+	UNIX_SOCKET_FILENAME = "autograder.sock"
+	STATUS_FILENAME      = "status.json"
+	UNIX_SOCKET_RANDNUM  = 32
+	PID_SOCK_LOCK        = "pid_sock_lock"
+)
+
+type StatusInfo struct {
+	Pid            int    `json:"pid"`
+	UnixSocketPath string `json:"unix_socket_path"`
+}
+
+func GetStatusPath() string {
+	return filepath.Join(config.GetWorkDir(), STATUS_FILENAME)
+}
+
+func WriteAndHandlePidStatus() error {
+	Lock(PID_SOCK_LOCK)
+	defer Unlock(PID_SOCK_LOCK)
+
+	statusPath := GetStatusPath()
+	pid := os.Getpid()
+	var statusJson StatusInfo
+
+	if !CheckAndHandlePid() {
+		return fmt.Errorf("Failed to check and handle the pid.")
+	}
+
+	statusJson.Pid = pid
+
+	err := util.ToJSONFile(statusJson, statusPath)
+	if err != nil {
+		return fmt.Errorf("Failed to write the pid.")
+	}
+
+	return nil
+}
+
+func GetUnixSocketPath() (string, error) {
+	Lock(PID_SOCK_LOCK)
+	defer Unlock(PID_SOCK_LOCK)
+
+	statusPath := GetStatusPath()
+	var statusJson StatusInfo
+
+	if util.IsFile(statusPath) {
+		err := util.JSONFromFile(statusPath, &statusJson)
+		if err != nil {
+			return "", fmt.Errorf("Failed to read the existing status file.")
+		}
+
+		if statusJson.UnixSocketPath != "" {
+			return statusJson.UnixSocketPath, nil
+		}
+	}
+
+	unixFileNumber, err := util.RandHex(UNIX_SOCKET_RANDNUM)
+	if err != nil {
+		return "", fmt.Errorf("Failed to generate a random number for the unix socket path.")
+	}
+
+	unixSocketPath := filepath.Join("/tmp", fmt.Sprintf("autograder-%s.sock", unixFileNumber))
+	statusJson.UnixSocketPath = unixSocketPath
+
+	err = util.ToJSONFile(statusJson, statusPath)
```

Comment:
I'm not a fan of the ordering here.
The file is created, and then this function comes in later to re-write the file with the socket path.

How about just creating the socket path when the stats file is first created.
We don't need to create the actual socket file,
so we can just generate the random number and reserve the path for later use.